frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -753,16 +672,36 @@ private void recoverTransitionTaskStateQueue() {
   private void addContainerRequest(TaskStatus taskStatus) throws Exception {
     String taskRoleName = taskStatus.getTaskRoleName();
     TaskStatusLocator taskLocator = new TaskStatusLocator(taskRoleName, taskStatus.getTaskIndex());
-    ContainerRequest request = setupContainerRequest(taskStatus);
+    Integer containerRequestTimeoutSec = CommonUtils.getRandomNumber(
+        conf.getLauncherConfig().getAmContainerRequestMinTimeoutSec(),
+        conf.getLauncherConfig().getAmContainerRequestMaxTimeoutSec());
 
-    LOGGER.logInfo("%s: addContainerRequest: %s", taskLocator, HadoopExtensions.toString(request));
+    ContainerRequest request = setupContainerRequest(taskStatus);
+    LOGGER.logInfo("%s: addContainerRequest with timeout %ss. ContainerRequest: [%s]",
+        taskLocator, containerRequestTimeoutSec, HadoopExtensions.toString(request));
     rmClient.addContainerRequest(request);
-    statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED);
+    statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED,
+        new TaskEvent().setContainerRequest(request));
+
+    transitionTaskStateQueue.queueSystemTaskDelayed(() -> {
+      if (statusManager.containsTask(request.getPriority())) {
+        LOGGER.logInfo(
+            "%s: ContainerRequest cannot be satisfied within timeout %ss, Cancel it and Request again. ContainerRequest: [%s]",
+            taskLocator, containerRequestTimeoutSec, HadoopExtensions.toString(request));
+        removeContainerRequest(taskStatus);
+        statusManager.transitionTaskState(taskLocator, TaskState.TASK_WAITING);
+        addContainerRequest(taskStatus);
```

Comment:
> addContainerRequest(taskStatus); [](start = 7, length = 33)

When time out happend, change the "priority" to the latest  sequential number can resolve this issue ?

Reply:
Already changed, pls check 

---
In reply to: [158204184](https://github.com/Microsoft/pai/pull/121#discussion_r158204184) [](ancestors = 158204184)